src/core/document.js

```diff
@@ -714,6 +687,67 @@ class PDFDocument {
     return shadow(this, "numPages", num);
   }
 
+  /**
+   * @private
+   */
+  _hasOnlyDocumentSignatures(fields, recursionDepth = 0) {
+    const RECURSION_LIMIT = 10;
+    return fields.every(field => {
+      field = this.xref.fetchIfRef(field);
+      if (field.has("Kids")) {
+        if (++recursionDepth > RECURSION_LIMIT) {
+          warn("_hasOnlyDocumentSignatures: maximum recursion depth reached");
+          return false;
+        }
+        return this._hasOnlyDocumentSignatures(
+          field.get("Kids"),
+          recursionDepth
+        );
+      }
+      const isSignature = isName(field.get("FT"), "Sig");
+      const rectangle = field.get("Rect");
+      const isInvisible =
+        Array.isArray(rectangle) && rectangle.every(value => value === 0);
+      return isSignature && isInvisible;
+    });
+  }
+
+  get formInfo() {
+    const formInfo = { hasAcroForm: false, hasXfa: false };
+    const acroForm = this.catalog.acroForm;
+    if (!acroForm) {
+      return shadow(this, "formInfo", formInfo);
+    }
+
+    try {
+      // The document contains XFA data if the `XFA` entry is a non-empty
+      // array or stream.
+      const xfa = acroForm.get("XFA");
+      const hasXfa = (Array.isArray(xfa) || isStream(xfa)) && xfa.length > 0;
```

Comment:
One final thing: When the `XFA`-entry is a Stream, can it ever be a something that's (indirectly) a `DecodeStream`?
In that case, there's no concept of exact Stream-length and consequently no `length` getter. Hence I believe that to prevent *any* future issues here, you want the following instead which is guaranteed to work for *all* Stream-implementations:
```suggestion
      const hasXfa =
        (Array.isArray(xfa) && xfa.length > 0) ||
        (isStream(xfa) && !xfa.isEmpty);
```

Reply:
The specification is not really clear on that, but since it wouldn't be the first time we see unexpected things in PDF files, I'll indeed handle this case. Moreover, the `isEmpty` getter is nicer since it works in all cases; I didn't know we had that.